appengine/src/test/java/com/google/googleinterns/gscribe/services/ExamServiceTest.java

```diff
@@ -0,0 +1,431 @@
+import com.google.googleinterns.gscribe.resources.io.exception.InvalidDatabaseDataException;
+import com.google.googleinterns.gscribe.resources.io.exception.InvalidRequestException;
+import com.google.googleinterns.gscribe.resources.io.request.ExamRequest;
+import com.google.googleinterns.gscribe.services.impl.ExamServiceImpl;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+import org.mockito.InjectMocks;
+import org.mockito.Mock;
+import org.mockito.junit.jupiter.MockitoExtension;
+
+import java.io.IOException;
+import java.security.GeneralSecurityException;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.mockito.Mockito.when;
+
+@ExtendWith(MockitoExtension.class)
+public class ExamServiceTest {
+
+    private final static User user = new User("user", "access_token", "refresh_token", null);
+    private final static String spreadsheetId = "spreadsheet_id";
+    private final static String sheetName = "sheet_name";
+    private final static ExamRequest examRequest = new ExamRequest();
+    private final static ValueRange valueRangeForNumberOfQuestions = new ValueRange();
+    private final static ValueRange valueRangeWhole = new ValueRange();
+    private final static List<List<Object>> globalValueRangeDataWhole = new ArrayList<>();
+
+    @Mock
+    SpreadsheetService spreadsheetService;
+    @InjectMocks
+    ExamServiceImpl examService;
+
+    @BeforeAll
+    private static void init() {
+        examRequest.setSheetName(sheetName);
+        examRequest.setSpreadsheetID(spreadsheetId);
+        List<List<Object>> valueRangeDataFoNumberOfQuestions = new ArrayList<>();
+        valueRangeDataFoNumberOfQuestions.add(Arrays.asList(new String[]{"Duration of exam:"}));
+        valueRangeDataFoNumberOfQuestions.add(Arrays.asList(new String[]{"Question Type"}));
+        valueRangeDataFoNumberOfQuestions.add(Arrays.asList(new String[]{"SUBJECTIVE"}));
+        valueRangeDataFoNumberOfQuestions.add(Arrays.asList(new String[]{"MCQ"}));
+        valueRangeForNumberOfQuestions.setValues(valueRangeDataFoNumberOfQuestions);
+        globalValueRangeDataWhole.add(Arrays.asList(new String[]{"Duration of exam:", "100"}));
+        globalValueRangeDataWhole.add(Arrays.asList(new String[]{"Question Type", "Question statement", "Option A", "Option B", "Option C", "Option D", "Points"}));
+        globalValueRangeDataWhole.add(Arrays.asList(new String[]{"SUBJECTIVE", "SubjectiveStatement1", "", "", "", "", "2"}));
+        globalValueRangeDataWhole.add(Arrays.asList(new String[]{"MCQ", "MCQStatement1", "OptionA", "OptionB", "OptionC", "OptionD", "3"}));
+    }
+
+    private List<List<Object>> globalValueRangeDataWholeCopy() {
+        List<List<Object>> copyList = new ArrayList<>();
```

Comment:
This is just creating a copy of the list, not the actual object. Try using clone().

Reply:
Using clone makes a shallow copy. So when this copyList is changed in the test for checking the functionality of test, the global list is getting changed too. The global list is set as an standard of what correct exam spreadsheet looks like. So I resorted to creating a new list every time needed so that any change to this new list won't affect the global standard list.

Reply:
Let's do a simple thing, create a separate function which returns valueRangeData, and don't store it in any global variable.

Reply:
Done.